packages/node/src/plugins/segmentio/publisher.ts

```diff
@@ -96,37 +117,48 @@ export class Publisher {
       and is always sent before a new batch is created.
 
       Add an event to the existing batch.
-        Success: Check if batch is full and send if it is.
+        Success: Check if batch is full or no more items are expected to come in (i.e. closing). If so, send batch.
         Failure: Assume event is too big to fit in current batch - send existing batch.
           Add an event to the new batch.
             Success: Check if batch is full and send if it is.
             Failure: Event exceeds maximum size (it will never fit), fail the event.
     */
-
-    if (batch.tryAdd(pendingItem)) {
-      if (batch.length === this._maxEventsInBatch) {
+    const addStatus = batch.tryAdd(pendingItem)
+    if (addStatus.success) {
+      const isExpectingNoMoreItems =
+        batch.length === this._closeAndFlushPendingItemsCount
+      const isFull = batch.length === this._maxEventsInBatch
+      if (isFull || isExpectingNoMoreItems) {
         this.send(batch).catch(noop)
         this.clearBatch()
       }
       return ctxPromise
-    } else if (batch.length) {
+    }
+
+    // If the new item causes the maximimum event size to be exceeded, send the current batch and create a new one.
+    if (batch.length) {
       this.send(batch).catch(noop)
       this.clearBatch()
     }
 
     const fallbackBatch = this.createBatch()
 
-    if (!fallbackBatch.tryAdd(pendingItem)) {
-      ctx.setFailedDelivery({
-        reason: new Error(`Event exceeds maximum event size of 32 kb`),
-      })
-      return Promise.resolve(ctx)
-    } else {
-      if (fallbackBatch.length === this._maxEventsInBatch) {
+    const fbAddStatus = fallbackBatch.tryAdd(pendingItem)
+
+    if (fbAddStatus.success) {
+      const isExpectingNoMoreItems =
+        fallbackBatch.length === this._closeAndFlushPendingItemsCount
```

Comment:
One thing I just thought of - wonder if we should decrement `_closeAndFlushPendingItemsCount` when we send a batch. Because if we had 10 events and only 5 fit in a batch, then we'll never hit the pending items count.

Reply:
That's an interesting scenario where batch size exceeds the pending item count. I will write a test for it.

Reply:
> That's an interesting scenario where batch size exceeds the pending item count.

Opposite actually - where max batch size is smaller than the pending item count. Imagine if batch size was 5 but you sent 6 events you'd always be expecting more currently :grin:

Reply:
You missed my edit!